README.md
# WA-Testing-Tool
Scripts that run against Watson Assistant for
  - `KFOLD` K fold cross validation on training set,
  - `BLIND` Evaluating a blind test, and
  - `TEST` Testing the WA against a list of utterances.

In the case of a k-fold cross validation, or a blind set, the tool will output
a [precision curve](precision_curve.md), in addition to per-intent precision and recall rates, and a confusion matrix.

## Features
- Easy to setup in one configuration file.
- Save the state when Assistant service is down in the middle of processing.
- Able to resume from where it stops using modularized scripts.

## Prerequisite
- Python 3.6.4 +
- Mac users: you may need to initialize Python's SSL certificate store by running `Install Certificates.command` found in `/Applications/Python`.  See more [here](https://github.com/cognitive-catalyst/WA-Testing-Tool/issues/38)
- Git client


## Quick Start
Pre-work: Make sure to cd into the location of a projects folder, where you will clone this github repo.  Within the folder, cd into the WA-Testing-Tool folder. 
1. Install code  `git clone https://github.com/cognitive-catalyst/WA-Testing-Tool.git`
2. Install dependencies `pip3 install --upgrade -r requirements.txt`
3. Set up parameters properly in configuration file (ex: `config.ini`). Use `config.ini.sample` to bootstrap your configuration.
  a. In your terminal, copy the config file into a new one, `cp config.ini.sample config.ini`
  b. Open the config.ini file in your favorite text editor, edit and save the following information with your actual credentials: 
      API Key
      url
      workspace_id (Watson Assistant v1) or environment_id (Watson Assistant v2)
  c. Set the mode and the mode-specific parameters.
4. Run the process. `python3 run.py -c config.ini` or `python3 run.py -c <path to your config file>`

## Quick Update
If you have already installed this utility use these steps to get the latest code.
1. Upgrade dependencies `pip3 install --upgrade -r requirements.txt`
2. Update to latest code level `git pull`

## Input Files
`config.ini` - Configuration file for `run.py`.
This is formatted differently for each mode.  Review the Examples below to explore the possible modes and how each is configured.

`test_input_file.csv` - Test set for blind testing and standard test.

For blind test with golden intent used for comparison:

| utterance            | golden intent                            |
| -------------------- | ---------------------------------------- |
| utterance 0          | intent 0                                 |
| utterance 1          | intent 0                                 |
| utterance 2          | intent 1                                 |

For standard test, the input must only have one column or error will be thrown:

| utterance            |
| -------------------- |
| utterance 0          |
| utterance 1          |
| utterance 2          |


## Examples
There are a variety of ways to use this tool.  Primarily you will execute a k-folds, blind, or standard test.
### Core execution modes
[Run k-fold cross-validation](examples/kfold.md)

[Run blind test](examples/blind.md)

[Run standard test without ground truth](examples/standard-test.md)

### Extended modes (executed by default)
[Generate precision/recall for classification test](examples/intent-metrics.md)

[Generate confusion matrix for classification test](examples/confusion-matrix.md)

[Compare two different blind test results](examples/compare-blind.md)

### Extended modes
[Generate description for intents](examples/intent-description.md)

[Generate long-tail classification results](examples/long-tail-scoring.md)

[Unit test dialog flows](dialog_test/README.md)

[Run syntax validation patterns on a workspace](validate_workspace/README.md)

[Extract and analyze Watson Assistant log data](log_analytics/README.md)

### More examples
Long-form resources available in Article and Video form:

Title | Article | Video
-- | -- | --
Testing a Chatbot with k-folds Cross Validation | https://medium.com/ibm-watson/testing-a-chatbot-with-k-folds-cross-validation-68dab111a6b | https://www.youtube.com/watch?v=FrhK68WyOK4
Analyze chatbot classifier performance from logs | https://medium.com/ibm-watson/analyze-chatbot-classifier-performance-from-logs-e9cf2c7ca8fd | https://www.youtube.com/watch?v=yd89DKyf6hc
Improve a chatbot classifier with production data | https://medium.com/ibm-watson/improve-a-chatbot-classifier-with-production-data-22a437f419b4 | https://www.youtube.com/watch?v=ftFIQtHiQY8

### Related projects
Watson Assistant is commonly paired with IBM Speech services to build voice-driven Conversational AI solutions.  Check out these tools to assess and tune your speech models!

* [STT-WER-Python](https://github.com/IBM/watson-stt-wer-python/): Utilities for testing IBM Speech to Text
* [TTS-Python](https://github.com/IBM/watson-tts-python/): Utilities for testing IBM Text to Speech

## Testing Natural Language Understanding Classifier
This tool can also be used to test a trained Natural Language Understanding (NLU) Classifier. The configuration is similar to testing Watson Assistant except:
1. Use the NLU URL in the `url` parameter (ex: `https://api.us-south.natural-language-understanding.watson.cloud.ibm.com`)
2. Specify the `<model_id>` in the `workspace_id` parameter in the configuration
3. Since NLU classifier does not support downloading training data, the original training data must be provided if run in 'kfold' mode (using the `train_input_file` parameter)

This NLU mode requires Python 3.9+ or later.

## General Caveats and Troubleshooting
1. Due to different coverage among service plans, user may need to adjust `max_test_rate` accordingly to avoid network connection error.

2. Users on Lite plans are only able to create 5 workspaces.  They should set `fold_num=3` on their k-fold configuration file.

3. In case of interrupted execution, the tool may not be able to clean up the workspaces it creates.  In this case you will need to manually delete the extra workspaces.

4. Workspace ID is *not* the Skill ID.  In the Watson Assistant user interface, the Workspace ID can be found on the Skills tab, clicking the three dots (top-right of skill), and choosing View API Details.

5. SSL: [CERTIFICATE_VERIFY_FAILED] on Mac means you may need to initialize Python's SSL certificate store by running `Install Certificates.command` found in `/Applications/Python`.  See more [here](https://github.com/cognitive-catalyst/WA-Testing-Tool/issues/38)

6. "This utility used to work and now it doesn't." Upgrade to latest dependencies with `pip3 install --upgrade -r requirements.txt` and latest code with `git pull`.

7. If you get a Python module loading error, confirm that you are using matching pip and python version, ie `pip3` and `python3` or `pip` and `python`.

8. Watson Assistant v2 configuration does not support k-folds mode.  Watson Assistant v2 is tested "in-place" rather than creating temporary skills for this tool.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## 2022-01-17
### Added
- Support for Watson Assistant v2 API log extraction and analysis

## 2022-12-15
### Added
- Support for Watson Assistant v2 API accuracy (blind test and standard test only)

## 2022-03-03
### Added
- Optionally compare two separate blind test runs and compute comparison metrics with `compare_results.py`

## 2020-11-12
### Changed
- Pass Watson Assistant system_settings to test workspaces

## 2020-04-10
### Changed
- validateWS.py is now moved to `log_analytics` folder.
- `waObjects.py` is extracted from that module to provide workspace parsing logic, usable in other modules including analytics

## 2020-01-15
### Added
- [Log extraction and basic analysis](log_analytics/README.md) capabilities in `log_analytics` folder.

## 2019-12-02
### Added
- Configurable k-fold union output file name

## 2019-11-15
### Added
- Support for testing Natural Language Classifier models

## 2019-10-28
### Removed
- Support for username/password authentication. (Watson APIs require API key authentication)

## 2019-09-30
### Added
- Generates confusion matrix in k-fold and blind modes

## 2019-09-26
### Added
- Default values for most configuration values.  Configuration effort is greatly reduced.

## 2019-09-23
### Added
- Utility for extracting conversation logs and utterances based on a given dialog node.

## 2019-09-03
### Updated
- [Static analysis utilities](validate_workspace/README.md) for conversation dialog evaluation.

## 2019-07-10
### Added
- Intent metrics include F1 score.

## 2019-06-21
### Added
- Support for IAM API key

## 2019-06-20
### Added
- [Static analysis utilities](validate_workspace/README.md) for conversation dialog evaluation.

## 2018-07-09
### Added
- Support optional partial credit intent tables

## 2018-07-03
### Added
- Support local workspace JSON as training input

## 2018-06-25
### Added
- Added intent metrics generation script

## 2018-06-14
### Added
- Made service base URL configurable

## 2018-05-21
### Added
- Refactored training input to only use workspace ID

## 2018-05-09
### Added
- Test cases for three modes and all sub-scripts are provided along with testing resources

## 2018-05-04
### Added
- Handling exception by cleaning out workspaces optionally after training phase.
- Exposed maximum test rate and weight mode parameters.

## 2018-04-30
### Fixed
- Allowed intents with confidence less than 0.2 to be returned.

### Added
- Calculate precision using different weighting configuration.
- Highlight confidence threshold on the precision curve figure.
- Union all folds test output as `kfold-test-out-union.csv`.

## 2018-04-13
### Added
- Intent description generation script is provided in [subfolder](intent-description).

## 2018-04-13
### Fixed
- Reserved entities can be imported to workspaces.

## 2018-04-12
### Added
- Coroutine-based concurrent testing is provided in sub-script.


utils/__init__.py
# coding: utf-8

# Copyright 2018 IBM All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import logging
import csv
import os
import pandas as pd
from ibm_watson import AssistantV1
from ibm_watson import NaturalLanguageUnderstandingV1
from ibm_cloud_sdk_core.authenticators import IAMAuthenticator, BearerTokenAuthenticator

UTF_8 = 'utf-8'
TRAIN_FILENAME = 'train.csv'
TEST_FILENAME = 'test.csv'
TEST_OUT_FILENAME = 'test-out.csv'
TRAIN_INTENT_FILENAME = 'intent-train.csv'
TRAIN_ENTITY_FILENAME = 'entity-train.csv'

# CSV column names
INTENT_COLUMN = 'intent'
CONFIDENCE_COLUMN = 'confidence'
UTTERANCE_COLUMN = 'utterance'  # used in any intermediate file
INTENT_JUDGE_COLUMN = 'does intent match'
PREDICTED_INTENT_COLUMN = 'predicted intent'
DETECTED_ENTITY_COLUMN = 'detected entity'
DIALOG_RESPONSE_COLUMN = 'dialog response'
GOLDEN_INTENT_COLUMN = 'golden intent'
SCORE_COLUMN = 'score'

WCS_IAM_APIKEY_ITEM = 'iam_apikey'
WCS_AUTH_TYPE_ITEM = 'auth_type'
WCS_BASEURL_ITEM = 'url'
WCS_CREDS_SECTION = 'ASSISTANT CREDENTIALS'
WA_API_VERSION_ITEM = 'version'
WA_DISABLE_SSL = 'disable_ssl'

SPEC_FILENAME = 'workspace.json'
WORKSPACE_BASE_FILENAME = 'workspace_base.json'

root_path = os.path.abspath(os.path.dirname(__file__))
# Sub-script paths
CREATE_TEST_TRAIN_FOLDS_PATH = os.path.join(root_path,
                                            'createTestTrainFolds.py')
TRAIN_CONVERSATION_PATH = os.path.join(root_path,
                                       'trainConversation.py')
TRAIN_CLASSIFIER_PATH = os.path.join(root_path,
                                       'trainNLC.py')
TEST_CONVERSATION_PATH = os.path.join(root_path,
                                      'testConversation.py')
TEST_CLASSIFIER_PATH = os.path.join(root_path,
                                      'testNLC.py')
CREATE_PRECISION_CURVE_PATH = os.path.join(root_path,
                                           'createPrecisionCurve.py')
WORKSPACE_PARSER_PATH = os.path.join(root_path,
                                     'workspaceParser.py')
INTENT_METRICS_PATH = os.path.join(root_path, 'intentmetrics.py')
CONFUSION_MATRIX_PATH = os.path.join(root_path, 'confusionmatrix.py')

# MODE
KFOLD = 'kfold'
BLIND_TEST = 'blind'
STANDARD_TEST = 'test'

FOLD_NUM_DEFAULT = 5
DEFAULT_WA_VERSION = '2021-06-14'
WORKSPACE_ID_TAG = 'workspace_id'
CLASSIFIER_ID_TAG = 'model_id'
ENVIRONMENT_ID_TAG = 'environment_id'
TIME_TO_WAIT = 600
BOOL_MAP = {True: 'yes', False: 'no'}
DEFAULT_TEST_RATE = 100
DEFAULT_CONF_THRES = 0.2
DEFAULT_TEMP_DIR = "./data"

POPULATION_WEIGHT_MODE = 'population'
EQUAL_WEIGHT_MODE = 'equal'

# WA BASE_URL - default if no URL is passed
BASE_URL = 'https://gateway.watsonplatform.net/assistant/api'

logger = logging.getLogger(__name__)


def configure_logger(level, format):
    """ Styling the logs in each module
    """
    logger.setLevel(level)
    h = logging.StreamHandler()
    h.setFormatter(logging.Formatter(format))
    logger.addHandler(h)


def save_dataframe_as_csv(df, file):
    """ Save dataframe as csv
    """
    return df.to_csv(file, encoding=UTF_8, quoting=csv.QUOTE_ALL,
                     index=False)


def marshall_entity(entities):
    """ Serialize RuntimeEntity list into formatted string
    """
    return ';'.join(['{}:{}'.format(entity['entity'], entity['value'])
                     for entity in entities])


def unmarshall_entity(entities_str):
    """ Deserialize entities string to RuntimeEntity list
    """
    entities = []
    for entity_str in entities_str.split(';'):
        splitted = entity_str.split(':')
        entities.append({'entity': splitted[0], 'value': splitted[0]})
    return entities


def delete_workspaces(iam_apikey, url, version, workspace_ids, auth_type, disable_ssl):
    """ Delete workspaces
    """
    if auth_type == 'iam':
        authenticator = IAMAuthenticator(iam_apikey)
    elif auth_type == 'bearer':
        authenticator = BearerTokenAuthenticator(iam_apikey)
    else:
        raise ValueError(f'Unknown auth_type "{auth_type}"')

    for workspace_id in workspace_ids:
        if 'natural-language-understanding' in url:
            c = NaturalLanguageUnderstandingV1(
                    version=version,
                    authenticator=authenticator
                    )
            c.set_service_url(url)
            c.delete_classifications_model(model_id=workspace_id)
        else:
            c = AssistantV1(
                    version=version,
                    authenticator=authenticator
                    )
            c.set_disable_ssl_verification(eval(disable_ssl))
            c.set_service_url(url)
            c.delete_workspace(workspace_id=workspace_id)

    print('Cleaned up workspaces')

def parse_partial_credit_table(file):
    df = pd.read_csv(file, quoting=csv.QUOTE_ALL,
                     encoding=UTF_8)
    df.fillna('0.0')
    table = {}
    for _, row in df.iterrows():
        golden_intent = row['Golden Intent'].strip()
        if golden_intent not in table:
            table[golden_intent] = {}
        table[golden_intent][row['Partial Credit Intent'].strip()] = \
            float(row['Partial Credit Intent Score'])

    return table
